File: app/src/main/java/com/abdotareq/subwaye_ticketting/ui/activity/SignInActivity.java
package com.abdotareq.subwaye_ticketting.ui.activity;

import androidx.appcompat.app.AppCompatActivity;

import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.abdotareq.subwaye_ticketting.R;
import com.abdotareq.subwaye_ticketting.databinding.ActivitySignInBinding;
import com.abdotareq.subwaye_ticketting.model.dto.Token;
import com.abdotareq.subwaye_ticketting.model.dto.User;
import com.abdotareq.subwaye_ticketting.model.retrofit.UserService;
import com.abdotareq.subwaye_ticketting.utility.SharedPreferenceUtil;
import com.abdotareq.subwaye_ticketting.utility.util;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class SignInActivity extends AppCompatActivity {

    private Retrofit retrofit;
    private ActivitySignInBinding binding;

    private Button signInBtn;
    private TextView recoverPassTv, signUpTv;
    private EditText mailEt, passEt;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sign_in);

        //initialize retrofit object
        retrofit = new Retrofit.Builder()
                .baseUrl(getString(R.string.base_url))
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        binding = ActivitySignInBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        setContentView(view);

        callListeners();

    }

    // Call listeners on the activity for code readability
    private void callListeners() {

        signInBtn = binding.signInBtn;
        signUpTv = binding.signInSignUpTv;
        recoverPassTv = binding.signInRecoverPassTv;

        signInBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                signInBtnClick();
            }
        });

        signUpTv.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(SignInActivity.this, SignUpActivity.class);
                startActivity(intent);
            }
        });

        recoverPassTv.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(SignInActivity.this, ForgetPassActivity.class);
                startActivity(intent);
            }
        });


    }

    // Handle sign in Button clicks
    private void signInBtnClick() {

        mailEt = binding.signInMailEt;
        passEt = binding.signInPassEt;

        mailEt.setText("abdo.elbishihi@gmail.com");
//        passEt.setText("abdo1234");

        //check for all inputs from user are not empty
        if (!util.isValidEmail(mailEt.getText().toString())) {
            mailEt.setText("");
            mailEt.setHint(getString(R.string.invalid_mail_warning));
            return;
        } else if (!util.isValidPassword(passEt.getText().toString())) {
            passEt.setText("");
            passEt.setHint(getString(R.string.pass_warning));
            return;
        }

        User user = new User(mailEt.getText().toString());

        user.setPassword(passEt.getText().toString());

        Log.e("SignInActivity: ", user.toString());

        authenticate(user);

    }

    /**
     * A method used to authenticate user (sign in)
     * can't be boolean as it has a thread which method won't wait until it finishes
     */
    private void authenticate(User user) {

        //create UserService object
        UserService service = retrofit.create(UserService.class);

        //initialize the save user call
        Call<Token> call = service.authenticate(user);

        //initialize and show a progress dialog to the user
        final ProgressDialog progressDialog = util.initProgress(this, getString(R.string.progMessage));
        progressDialog.show();

        //start the call
        call.enqueue(new Callback<Token>() {
            @Override
            public void onResponse(Call<Token> call, Response<Token> response) {

                int responseCode = response.code();

                if (responseCode >= 200 && responseCode <= 299 && response.body() != null) {
                    //user authenticated successfully
                    progressDialog.dismiss();

                    //write token into SharedPreferences
                    SharedPreferenceUtil.setSharedPrefsLoggedIn(SignInActivity.this, true);
                    SharedPreferenceUtil.setSharedPrefsUserId(SignInActivity.this, response.body().getToken());

                    Intent intent =new Intent(SignInActivity.this,HomeLandActivity.class);
                    startActivity(intent);

                    passEt.setHint("token:   " + response.body().getToken());

//                        if (getIntent().hasExtra("LOGGED_OUT")) {
//                            Intent mainIntent = new Intent(SignInActivity.this, MainActivity.class);
//                            startActivity(mainIntent);
//                        }
//                    Intent intent = new Intent(SignInActivity.this, MainActivity.class);
//                    startActivity(intent);
//                    finish();
                } else if (responseCode == 436) {
                    //user not authenticated successfully
                    progressDialog.dismiss();
                    Toast.makeText(SignInActivity.this, getText(R.string.wrong_mail_or_pass), Toast.LENGTH_LONG).show();
                } else {
                    //user not authenticated successfully
                    progressDialog.dismiss();
                    Toast.makeText(SignInActivity.this, "else onResponse", Toast.LENGTH_LONG).show();
                }
            }

            @Override
            public void onFailure(Call<Token> call, Throwable t) {

                progressDialog.dismiss();
                Toast.makeText(SignInActivity.this, getText(R.string.error_message), Toast.LENGTH_LONG).show();
            }
        });


    }


}

File: app/src/main/java/com/abdotareq/subway_e_ticketing/controller/activity/registration/SignInActivity.kt
package com.abdotareq.subway_e_ticketing.controller.activity.registration

import android.content.Intent
import android.os.Bundle
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.abdotareq.subway_e_ticketing.R
import com.abdotareq.subway_e_ticketing.databinding.ActivitySignInBinding
import com.abdotareq.subway_e_ticketing.model.dto.Token
import com.abdotareq.subway_e_ticketing.model.dto.User
import com.abdotareq.subway_e_ticketing.model.retrofit.UserApiObj
import com.abdotareq.subway_e_ticketing.controller.activity.HomeLandActivity
import com.abdotareq.subway_e_ticketing.utility.SharedPreferenceUtil
import com.abdotareq.subway_e_ticketing.utility.util
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import timber.log.Timber

/**
 * The Activity Controller Class that is responsible for handling Signing in
 */
class SignInActivity : AppCompatActivity() {

    private lateinit var binding: ActivitySignInBinding
    private lateinit var passEt: EditText

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sign_in)

        binding = ActivitySignInBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)
        callListeners()
    }

    // Call listeners on the activity for code readability
    private fun callListeners() {
        val signInBtn = binding.signInBtn
        val signUpTv = binding.signInSignUpTv
        val recoverPassTv = binding.signInRecoverPassTv
        signInBtn.setOnClickListener { signInBtnClick() }
        signUpTv.setOnClickListener {
            val intent = Intent(this@SignInActivity, SignUpActivity::class.java)
            startActivity(intent)
        }
        recoverPassTv.setOnClickListener {
            val intent = Intent(this@SignInActivity, RecoverPassActivity::class.java)
            startActivity(intent)
        }
    }

    // Handle sign in Button clicks
    private fun signInBtnClick() {
        val mailEt = binding.signInMailEt
        passEt = binding.signInPassEt

        mailEt.setText("abdo.elbishihi@gmail.com")
        passEt.setText("abdo1234")
        Toast.makeText(this, "Const values written in signInBtnClick method ", Toast.LENGTH_LONG).show()

        //check for all inputs from user are not empty
        if (util.isValidEmail(mailEt.text.toString())) {
            mailEt.setText("")
            mailEt.hint = getString(R.string.invalid_mail_warning)
            return
        } else if (!util.isValidPassword(passEt.text.toString())) {
            passEt.setText("")
            passEt.hint = getString(R.string.pass_warning)
            return
        }
        val user = User()
        user.email = mailEt.text.toString()
        user.password = passEt.text.toString()
        Timber.e(user.toString())

        authenticate(user)
    }

    /**
     * A method used to authenticate user (sign in)
     * can't be boolean as it has a thread which method won't wait until it finishes
     */
    private fun authenticate(user: User) {

        //initialize and show a progress dialog to the user
        val progressDialog = util.initProgress(this, getString(R.string.progMessage))
        progressDialog.show()

        //start the call
        UserApiObj.retrofitService.authenticate(user)?.enqueue(object : Callback<Token?> {
            override fun onResponse(call: Call<Token?>, response: Response<Token?>) {
                val responseCode = response.code()
                if (responseCode in 200..299 && response.body() != null) {
                    //user authenticated successfully
                    progressDialog.dismiss()

                    //write token into SharedPreferences to use in remember user
                    SharedPreferenceUtil.setSharedPrefsLoggedIn(this@SignInActivity, true)
                    SharedPreferenceUtil.setSharedPrefsTokenId(this@SignInActivity, response.body()!!.token)
                    Timber.e("token:    ${response.body()!!.token}")
                    val intent = Intent(this@SignInActivity, HomeLandActivity::class.java)
                    startActivity(intent)
                    finish()

                } else if (responseCode == 436) {
                    //user not authenticated successfully
                    progressDialog.dismiss()
                    Toast.makeText(this@SignInActivity, getText(R.string.wrong_mail_or_pass), Toast.LENGTH_LONG).show()
                } else {
                    //user not authenticated successfully
                    progressDialog.dismiss()
                    Toast.makeText(this@SignInActivity, "else onResponse", Toast.LENGTH_LONG).show()
                }
            }

            override fun onFailure(call: Call<Token?>, t: Throwable) {
                progressDialog.dismiss()
                Timber.e("getText(R.string.error_message)${t.message}")
                Toast.makeText(this@SignInActivity, getText(R.string.failure_happened), Toast.LENGTH_LONG).show()
            }
        })
    }
}